pkg/resources/opnicluster/services.go

```diff
@@ -600,19 +607,192 @@ func (r *Reconciler) gpuCtrlDeployment() (runtime.Object, reconciler.DesiredStat
 	return deployment, deploymentState(r.opniCluster.Spec.Services.GPUController.Enabled), nil
 }
 
+func (r *Reconciler) getPrometheusEndpoint() (endpoint string) {
+	lg := log.FromContext(r.ctx)
+	if r.opniCluster.Spec.Services.Metrics.PrometheusReference != nil {
+		prometheus := &monitoringv1.Prometheus{}
+		err := r.client.Get(r.ctx, types.NamespacedName{
+			Name:      r.opniCluster.Spec.Services.Metrics.PrometheusReference.Name,
+			Namespace: r.opniCluster.Spec.Services.Metrics.PrometheusReference.Namespace,
+		}, prometheus)
+		if err != nil {
+			lg.V(1).Error(err, "unable to fetch prometheus")
+		} else {
+			endpoint = prometheus.Spec.ExternalURL
+		}
+	}
+	if r.opniCluster.Spec.Services.Metrics.PrometheusEndpoint != "" {
+		endpoint = r.opniCluster.Spec.Services.Metrics.PrometheusEndpoint
+	}
+
+	return
+}
+
 func (r *Reconciler) metricsDeployment() (runtime.Object, reconciler.DesiredState, error) {
 	deployment := r.genericDeployment(v1beta1.MetricsService)
-	_, err := url.ParseRequestURI(r.opniCluster.Spec.Services.Metrics.PrometheusEndpoint)
+	prometheusEndpoint := r.getPrometheusEndpoint()
+	_, err := url.ParseRequestURI(prometheusEndpoint)
 	if err != nil && (r.opniCluster.Spec.Services.Metrics.Enabled == nil || *r.opniCluster.Spec.Services.Metrics.Enabled) {
 		return deployment, deploymentState(r.opniCluster.Spec.Services.Metrics.Enabled), errors.New("prometheus endpoint is not a valid URL")
 	}
 	deployment.Spec.Template.Spec.Containers[0].Env = append(deployment.Spec.Template.Spec.Containers[0].Env, corev1.EnvVar{
 		Name:  "PROMETHEUS_ENDPOINT",
-		Value: r.opniCluster.Spec.Services.Metrics.PrometheusEndpoint,
+		Value: prometheusEndpoint,
 	})
 	return deployment, deploymentState(r.opniCluster.Spec.Services.Metrics.Enabled), nil
 }
 
+func (r *Reconciler) metricsService() (runtime.Object, reconciler.DesiredState, error) {
+	labels := r.serviceLabels(v1beta1.MetricsService)
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      labels[resources.AppNameLabel],
+			Namespace: r.opniCluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
```

Comment:
Instead of setting this manually, we should use `ctrl.SetControllerReference()` which also sets the Controller field in the owner reference (also applies to other usages in this file)

Reply:
We don't use this for any of the other resources so I've stuck to what we are already using.  If we want to change this we should do it for all to keep them consistent

Reply:
We may be setting them incorrectly in some places - SetControllerReference sets `Controller: true` in the owner reference. If that field is not set I believe it will prevent things like finalizers from running and it also affects deletion propagation.